api/src/Auth/Providers/JWTAuthDriverServiceProvider.php

```diff
@@ -0,0 +1,180 @@
+            $model = $app['config']['auth.model'];
+            return new UserProvider($app['hash'], $model, $this->getTokenUserProvider());
+        });
+    }
+
+    protected function registerRequestParser()
+    {
+        $this->app->bind(RequestParser::class, function($app){
+            return new RequestParser(
+                $this->requestMethod,
+                $this->requestHeader,
+                $this->requestQuery,
+                $this->requestCookie
+            );
+        });
+    }
+
+    protected function registerDriver()
+    {
+        $this->app->extend('auth', function(AuthManager $auth, $app){
+            return $auth->extend('spira', function($app) {
+                return $app[Guard::class];
+            });
+        });
+
+        //extra rebinding for tests
+        /** @var Application $app */
+        $app = $this->app;
+        $app->rebinding(Request::class, function(Application $app, Request $request){
+            if ($app->resolved('auth')){
+                $app['auth']->setRequest($request);
+            }
+        });
+    }
+
+    protected function registerTokenizer()
+    {
+        $this->app->bind(JWTInterface::class, function($app){
+            return new NamshiAdapter($this->getSecretPublic(), $this->getSecretPrivate(), $this->algorithm);
+        });
+    }
+
+    protected function registerPayloadFactory()
+    {
+        $this->app->bind(PayloadFactory::class, function($app){
+            return new PayloadFactory($this->getPayloadGenerators());
+        });
+    }
+
+    protected function registerPayloadValidatorFactory()
+    {
+        $this->app->bind(PayloadValidationFactory::class, function($app){
+            return new PayloadValidationFactory($this->getValidationRules());
+        });
+    }
+
+    protected function getTokenUserProvider()
+    {
+        return null;
+    }
+
+
+    /**
+     * iis - the Issuer claim
+     * iat - Issued At claim
+     * exp - Expired At claim
+     * nbf - Not Before claim
+     * jti - token unique id
+     *
+     * @return array
+     */
+    protected function getPayloadGenerators()
+    {
+        /** @var Request $request */
+        $request = $this->app['request'];
+        return [
+            'iss'=> function() use ($request) { return $request->url();},
+            'iat'=> function(){ return Carbon::now()->format('U');},
+            'exp'=> function(){ return Carbon::now()->addMinutes($this->ttl)->format('U');},
+            'nbf'=> function(){ return Carbon::now()->format('U');},
+            'jti'=> function(){ return str_random(16);},
+        ];
+    }
+
+    protected function getValidationRules()
+    {
+        return [
+            'nbf' => function($payload){
+                if ( Carbon::createFromTimeStampUTC($payload['nbf'])->isFuture()) {
+                    throw new TokenInvalidException('Not Before (nbf) timestamp cannot be in the future', 400);
```

Comment:
Use constants for the response codes, ideally have it defaulted in the exception class anyway.